vesc_hw_interface/src/vesc_servo_controller.cpp

```diff
@@ -74,60 +81,89 @@ void VescServoController::init(ros::NodeHandle nh, VescInterface* interface_ptr)
   {
     ROS_ERROR("[Servo Calibration] Invalid mode");
   }
+
+  // Smoothing differentiation when hall sensor resolution is insufficient
+  bool smooth_diff;
+  nh.param<bool>("servo/enable_smooth_diff", smooth_diff, true);
+  if (smooth_diff)
+  {
+    double smooth_diff_max_sampling_time;
+    int counter_td_vw_max_step;
+    nh.param<double>("servo/smooth_diff/max_sample_sec", smooth_diff_max_sampling_time, 1.0);
+    nh.param<int>("servo/smooth_diff/max_smooth_step", counter_td_vw_max_step, 10);
+    vesc_step_difference_.enableSmooth(control_rate_, smooth_diff_max_sampling_time, counter_td_vw_max_step);
+    ROS_INFO("[Motor Control] Smooth differentiation enabled, max_sample_sec: %f, max_smooth_step: %d",
+             smooth_diff_max_sampling_time, counter_td_vw_max_step);
+  }
   // Create timer callback for PID servo control
-  control_timer_ = nh.createTimer(ros::Duration(control_period_), &VescServoController::controlTimerCallback, this);
+  control_timer_ = nh.createTimer(ros::Duration(1.0 / control_rate_), &VescServoController::controlTimerCallback, this);
   return;
 }
 
-void VescServoController::control(const double position_reference, const double position_current)
+void VescServoController::control()
 {
   // executes calibration
   if (calibration_flag_)
   {
-    calibrate(position_current);
+    calibrate(sens_pose_);
     // initializes/resets control variables
-    time_previous_ = ros::Time::now();
-    position_sens_previous_ = position_current;
-    position_reference_previous_ = calibration_position_;
-    error_previous_ = 0.0;
+    sens_pose_previous_ = sens_pose_;
+    target_pose_previous_ = calibration_position_;
     return;
   }
 
-  const ros::Time time_current = ros::Time::now();
-  // calculates PD control
-  const double error_current = position_reference - position_current;
-  const double u_pd = Kp_ * error_current + Kd_ * (error_current - error_previous_) / control_period_;
+  // calculates PID control
+  double step_diff = vesc_step_difference_.getStepDifference(position_steps_, false);
+  double current_vel = step_diff * 2.0 * M_PI / (num_rotor_poles_ * num_hall_sensors_) * control_rate_ * gear_ratio_;
+  double target_vel = (target_pose_limited_ - target_pose_previous_) / control_rate_;
 
-  double u = 0.0;
-
-  // calculates I control if PD input is not saturated
-  if (isSaturated(u_pd))
+  double error = target_pose_limited_ - sens_pose_;
+  double error_dt = target_vel - current_vel;
```

Comment:
Isn't `error_dt` the change in error? Is there are reason why we are using the difference between the current and target velocity converted from a position component?

Reply:
You are correct, error_dt is the change in error.
I think the current implementation is simpler than creating a new error_prev_ variable, but is there a question somewhere?

Reply:
My bad, I should had been a little more specific.

If I'm remember correctly, `error_dt = target_vel - current_vel` and `error_dt = (error - prev_error) /dt` is the same thing if the velocity is calculated such that  `target_velocity = (target_position - prev_position)/dt`. 

However `current_vel` is calculated differently if smoothing is enabled (If smoothing is disabled, VescStepDifference takes the previous steps which is essentially the same thing as above). It's essentially sensing the current velocity through tachometer steps.

I assume the error is taken from one source of sensor which, for this case, is the tachometer steps. While the `current_vel` with smoothing enable uses tachometer steps, is it not practically estimating current velocity and not the change in error?

Reply:
Yes, getStepDifference is used to estimate the current speed.
The target speed is not changed, but the current speed is smoothed.
This means that it slows down the response of the D control to the measured speed while not slowing down the response of the D control to the target speed.

Reply:
I thought we are using D gain to control the change in error, not the the measured (smoothed) speed? In addition, the deriving the position using smoothing method doesn't seem guarantee that all the queue data would be considered?

Reply:
D-gain is used to control the change in time differentiation of position error.
Differentiating the position error by time yields the difference between the target and measured velocities.
The target velocity is obtained by differentiating the target position with time.
Smoothing is used to derive measured velocity.
The smoothing method is not used to derive position.
If you are pointing out that this is different from normal PID control, you are correct.
This process is the same as lowering the D-gain when the target velocity is constant and the measured velocity is small.

Reply:
In practice, the calculation is equivalent to lowering the D gain and adding a feedforward proportional to the time derivative of the target angle.

Reply:
Ok make sense, I don't have any problem so far with how `error_dt` is calculated. Will leave this be.